### ISSUE TITLE
Failed to expire automount, ignoring: No such device

### ISSUE THREAD

**Issue by evgkrsk** [systemd/systemd#6776]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> 234

### Used distribution

> ALT Linux Sisyphus

Here is small part of my laptop's journal:
```
2017-09-08T20:27:07.022827+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-08T20:27:07.142875+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Mount process exited, code=exited status=16
2017-09-08T20:27:07.143830+07:00 thinkpad systemd[1]: Failed unmounting NFS share on nas:/music.
2017-09-08T20:27:07.144470+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device
2017-09-08T20:47:07.772585+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-08T20:47:07.870145+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Mount process exited, code=exited status=16
2017-09-08T20:47:07.871263+07:00 thinkpad systemd[1]: Failed unmounting NFS share on nas:/music.
2017-09-08T20:47:07.872033+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device
2017-09-08T21:07:08.524780+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-08T21:07:08.638970+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Mount process exited, code=exited status=16
2017-09-08T21:07:08.640148+07:00 thinkpad systemd[1]: Failed unmounting NFS share on nas:/music.
2017-09-08T21:07:08.640841+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device
2017-09-08T21:27:09.276683+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-08T21:27:09.378902+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Mount process exited, code=exited status=16
2017-09-08T21:27:09.379798+07:00 thinkpad systemd[1]: Failed unmounting NFS share on nas:/music.
2017-09-08T21:27:09.380562+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device
2017-09-08T21:36:56.380594+07:00 thinkpad systemd[1]: mnt-nas-homes.automount: Got automount request for /mnt/nas/homes, triggered by 8700 (firefox-bin)
2017-09-08T21:36:56.383589+07:00 thinkpad systemd[1]: Mounting NFS share on nas:/homes...
2017-09-08T21:36:56.483934+07:00 thinkpad systemd[1]: Mounted NFS share on nas:/homes.
```

Most of journal's records is just complains about failed unmounting of automount units. Is it way to make these units more quiet?

**Comment by poettering** [comment]
Well, something is really failing here, and we should not suppress log messages about real failures. Your umount command fails for some reason:

mnt-nas-music.mount: Mount process exited, code=exited status=16

The question is, why? According to the mount(8) man page the exit code 16 means "problems writing or locking /etc/mtab". The thing though is, we explicitly do not support systems where mtab is still enabled, see README. Please ping your distro regarding this issue, it appears they are settings things up incorrectly for systemd to work.

Either way, I am strictly against taping over these issues. We need to log about errors happening. And the right way to deal with errors is to fix them, not to disable log messages about them.

**Comment by evverx** [comment]
It seems that `umount` hasn't returned `16` since `deprecated-mount/umount.c` was replaced by `sys-utils/umount.c`. `deprecated-mount` was disabled in https://github.com/karelzak/util-linux/commit/434b07a09062f5c06b. Then it was completely removed in https://github.com/karelzak/util-linux/commit/65acdc8734535b7a.

Could you clarify what version of `umount` you are using?

+1 for closing the issue.

**Comment by evverx** [comment]
Given that [`MNT_EX_FILEIO`](https://github.com/karelzak/util-linux/blob/a8771daefecf8201cdb96e493e61b048d783842c/libmount/src/libmount.h.in#L230-L235) is never used, would it make sense to remove [the paragraph explaining what it is](https://github.com/karelzak/util-linux/blob/a8771daefecf8201cdb96e493e61b048d783842c/sys-utils/mount.8#L2986-L2988) from the man page? @karelzak what do you think?

**Comment by evgkrsk** [comment]
My util-linux version is 2.29.2

Seems like this annoying log messages present when I run mount triggering program (cmus, music player) via firejail (http://firejail.wordpress.com/). Without it, automatic unmounting works quietly (and successfully).

**Comment by evverx** [comment]
That's weird.

Is there any chance you could run the commands below
```sh
echo DefaultEnvironment=LIBMOUNT_DEBUG=all | sudo tee -a /etc/systemd/system.conf
sudo systemctl daemon-reload
sudo strace -p 1 -f -s500 -v -yy -o OUT &
```
, then trigger the error and paste the contents of `OUT` on `gist.github.com`?

**Comment by topimiettinen** [comment]
Firejail uses namespaces and bind mounts extensively. Perhaps these are done in a way that confuses systemd?

**Comment by evgkrsk** [comment]
@evverx, there is it: https://gist.github.com/evgkrsk/13a4c4bfc472a68bba520cccf0dec289

And there is corresponding syslog messages:
```
2017-09-10T14:59:39.485790+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-10T14:59:39.722612+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Mount process exited, code=exited status=16
2017-09-10T14:59:39.726253+07:00 thinkpad systemd[1]: Failed unmounting NFS share on nas:/music.
2017-09-10T14:59:39.727624+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device

2017-09-10T15:00:40.483977+07:00 thinkpad systemd[1]: Unmounting NFS share on nas:/music...
2017-09-10T15:00:40.786914+07:00 thinkpad systemd[1]: Unmounted NFS share on nas:/music.
2017-09-10T15:00:40.787933+07:00 thinkpad systemd[1]: Failed to expire automount, ignoring: No such device
2017-09-10T15:00:40.788908+07:00 thinkpad systemd[1]: mnt-nas-music.mount: Unit entered failed state.
```

**Comment by evverx** [comment]
Thank you. The error has nothing to do with `mtab`, because `16 (EBUSY)` is the value of `errno` which was set by `umount.nfs` having called `umount2("/mnt/nas/music", 0)`.

I'm not sure what exactly is going on here, but I'm certain that `systemd` shouldn't suppress logs.

**Comment by evverx** [comment]
Talking of suppressing logs, is there a reason for sending the output of `mount/umount` to `/dev/null`? I think the line
```
write(2</dev/null>, "umount.nfs: /mnt/nas/music: device is busy\n", 43) = 43
```
in the journal would a bit simplify the process of guessing what `16` is.

**Comment by poettering** [comment]
> Talking of suppressing logs, is there a reason for sending the output of mount/umount to /dev/null?

Yes, there is. We default to StandardError=null for mount units, as they should work before the journal is started. If we'd set StandardError=journal for them, they'd gain an automatic dep on systemd-journald.socket, and that might result in deadlocks if the mount is necessary to start the journal.

That said, for most mounts this is not even an issue, and we could relax this, question is how to detect if a mount unit is one that is required by the journal or not...

**Comment by poettering** [comment]
So, the question I guess is why is systemd trying to unmount this mount even though it apparently remains busy. @evgkrsk any chance you can try to reproduce this with systemd debug logging enabled? i.e. boot with systemd.log_level=debug and then paste a relevant excerpt of the logs here. This should tell us why systemd thinks it's a good idea to unmount the autofs share at that point in time.

**Comment by poettering** [comment]
Hmm, but indeed autofs might be confused by your mount namespaces, and doesn't "see" the accesses made through them. If that's the case, then there's little we can do about this in userspace, it's a kernel issue to fix.

**Comment by evgkrsk** [comment]
I switched from autofs4 to systemd's automount units some time ago. I haven't tried to run programs that triggering automount with autofs, but can check it.

I will try to run with systemd.log_level=debug and post journal content here...

**Comment by poettering** [comment]
sorry, i didn't mean "autofs" as referring to the original autofs automount userspace package, but simply to the kernel interface that is used by both systemd and that original package. No need to test with the original package.

**Comment by karelzak** [comment]
@evverx good catch. It's mistake that MNT_EX_FILEIO is not used. Fixed. Thanks!

**Comment by evgkrsk** [comment]
Here is output from "journalctl -b -1 -x" for boot with "systemd.log_level=debug".
[long.log.zip](https://github.com/systemd/systemd/files/1309502/long.log.zip)

**Comment by tim-seoss** [comment]
FWIW, I've seen this behaviour with Linux 5.16.12, systemd 247.3 and a combination of automount and bind mount unit files, when `ReadWritePaths=` (i.e. namespaces) are also in use. The behaviour looks like a kernel autofs bug. Please let me know if it would be helpful to put together a minimal reproducer.

**Comment by ajs124** [comment]
I've also been observing this with kernel 5.15.x and systemd 251.x.
My config also uses a service that has `ReadWritePaths` for a mountpoint, that gets auto unmounted via `x-systemd.automount` + `x-systemd.idle-timeout=30`. The umount seems successful, so idk why it would mark the unit as failed with this error.

**Comment by github-cygwin** [comment]
I can still reproduce this easily on kernel 7.1.8 with systemd 258 with a qemu instance running a Windows VM mounting a ISO image as CD/DVD from the automounted NFS share.

If no other process uses the automounted share at the time, systemd constantly tries to expire the automount based on the idle-timeout and fails with "No such device".

What's notable in this situation is that `fuser <automount>` does NOT print any PID.

However, `lsof <automount>` finds and prints the aforementioned qemu-system-x86_64 process.
